File: src/api/templates/teamPresets.ts
import { invokeWithTimeout as invoke } from "@/lib/tauriInvoke";
import { useI18nStore } from "@/stores/i18nStore";

import type { TeamPreset } from "@/lib/bindings/TeamPreset";
import type { AdoptedTeamPresetResult } from "@/lib/bindings/AdoptedTeamPresetResult";
import type { PresetAdoptionSchema } from "@/lib/bindings/PresetAdoptionSchema";

/**
 * Resolve the user's current locale to pass to the Rust preset loader.
 * Returns `null` for English so the backend short-circuits the overlay
 * lookup path entirely (canonical file IS English — no sibling needed).
 *
 * Read at call time (not module load) so a language switch mid-session
 * affects the next preset operation without a reload.
 */
function currentLanguage(): string | null {
  const lang = useI18nStore.getState().language;
  return lang === "en" ? null : lang;
}

// ============================================================================
// Team Presets — filesystem-shipped multi-template bundles
// ============================================================================
//
// A team preset is a manifest under `scripts/templates/_team_presets/*.json`
// that bundles N persona templates plus their team wiring (PersonaTeam
// metadata + members + connections, optionally a PersonaGroup binding).
// The Presets gallery (PresetLibraryPage) and the preview/adoption modal
// (PresetPreviewModal) consume these wrappers.
//
// Loading is read-fresh-from-disk per call — presets are tiny and rarely
// listed, so a stale cache trap (annoying during dev edits to the manifest)
// is worse than the few-ms read cost.

export const listTeamPresets = () =>
  invoke<TeamPreset[]>("list_team_presets", { language: currentLanguage() });

export const getTeamPreset = (id: string) =>
  invoke<TeamPreset>("get_team_preset", { id, language: currentLanguage() });

/**
 * Aggregate every member template's `payload.adoption_questions[]` for
 * the combined preset questionnaire. Returns one row per preset member
 * (members with no questions appear with an empty `questions` array
 * rather than being dropped, so the UI can show "no config needed").
 *
 * Failures to LOAD a member's template (renamed, deleted, parse error)
 * silently skip that row at the schema view — the same template will
 * raise a precise error when the adopter tries to use it. The
 * questionnaire screen treats schema-load best-effort; the adopt
 * action is the source of truth for member success/failure.
 */
export const getPresetAdoptionSchema = (presetId: string) =>
  invoke<PresetAdoptionSchema>("get_preset_adoption_schema", {
    presetId,
    language: currentLanguage(),
  });

/**
 * Run a preset's full adoption flow. Emits `team-preset-adopt-progress`
 * events per member transition (queued → adopting → done/failed) — wire
 * `useTypedTauriEvent(EventName.TEAM_PRESET_ADOPT_PROGRESS, ...)` to
 * drive per-row status badges in the preview modal.
 *
 * Returns an aggregate result with the new team_id, optional group_id,
 * successfully-adopted members, and any per-template failures. The team
 * shell is created unconditionally — partial failures leave the user
 * with the team + the members that did succeed so they can retry the
 * rest without losing progress.
 */
export const adoptTeamPreset = (id: string) =>
  invoke<AdoptedTeamPresetResult>("adopt_team_preset", {
    id,
    language: currentLanguage(),
  });

/**
 * Retry the named failed roles of a previously-adopted preset. Reuses
 * the same TEAM_PRESET_ADOPT_PROGRESS event stream so the modal's per-
 * row status badges animate identically to the first attempt.
 * Idempotent on roles already present in the team (silently skipped),
 * so double-clicking is safe.
 *
 * Returns the FULL member list (old + newly-retried) plus any roles
 * that still failed this round, so the modal can swap state in one
 * assignment instead of merging two views.
 */
export const retryTeamPresetMembers = (
  presetId: string,
  teamId: string,
  groupId: string | null,
  roles: string[],
) =>
  invoke<AdoptedTeamPresetResult>("retry_team_preset_members", {
    presetId,
    teamId,
    groupId,
    roles,
    language: currentLanguage(),
  });

File: src-tauri/src/commands/design/team_presets.rs
//! Tauri command surface for filesystem-shipped team presets.
//!
//! The `list_team_presets` / `get_team_preset` pair powers the Presets
//! gallery and the preview modal; `adopt_team_preset` will live next to
//! these once the adopter module lands. Loading is delegated entirely to
//! `engine::team_preset_loader` — these commands are auth-gated thin
//! wrappers.

use std::sync::Arc;
use tauri::{AppHandle, State};

use crate::db::models::{AdoptedTeamPresetResult, PresetAdoptionSchema, TeamPreset};
use crate::engine::{team_preset_adopter, team_preset_loader};
use crate::error::AppError;
use crate::ipc_auth::require_auth_sync;
use crate::AppState;

/// Return every valid preset manifest under
/// `scripts/templates/_team_presets/`. Invalid manifests are skipped
/// silently (logged via `tracing::warn` in the loader). Result is sorted
/// by `name` for stable gallery ordering.
///
/// `language` (e.g. `"zh"`, `"de"`) selects per-preset locale overlay
/// siblings (`<id>.<lang>.json`) when present. `None` returns canonical
/// English. Overlays are partial — fields absent from the overlay fall
/// through to canonical, so a translation team can ship name +
/// description first and add member-level translations later.
#[tauri::command]
pub fn list_team_presets(
    state: State<'_, Arc<AppState>>,
    language: Option<String>,
) -> Result<Vec<TeamPreset>, AppError> {
    require_auth_sync(&state)?;
    Ok(team_preset_loader::list_presets(language.as_deref()))
}

/// Return one preset by id (the on-disk filename minus `.json`). Returns
/// `NotFound` for an unknown id, `Validation` for parse/schema/role
/// failures.
///
/// `language` applies the matching `<id>.<lang>.json` overlay — same
/// rules as `list_team_presets`.
#[tauri::command]
pub fn get_team_preset(
    state: State<'_, Arc<AppState>>,
    id: String,
    language: Option<String>,
) -> Result<TeamPreset, AppError> {
    require_auth_sync(&state)?;
    team_preset_loader::get_preset(&id, language.as_deref())
}

/// Aggregate every member template's `payload.adoption_questions[]`
/// into one combined questionnaire schema for the Presets preview
/// modal. Members whose template files are missing or unparseable are
/// skipped from the schema view (the adopter surfaces those failures
/// at adopt time); members with no adoption_questions appear with an
/// empty `questions` array so the UI can render the full member list
/// rather than silently dropping rows.
///
/// `language` flows through to the preset-level metadata
/// (preset_name). Template-level translation of question labels is
/// applied frontend-side via the same template-overlay path the
/// single-template adoption flow uses, so callers don't need to
/// thread it twice.
#[tauri::command]
pub fn get_preset_adoption_schema(
    state: State<'_, Arc<AppState>>,
    preset_id: String,
    language: Option<String>,
) -> Result<PresetAdoptionSchema, AppError> {
    require_auth_sync(&state)?;
    team_preset_loader::get_adoption_schema(&preset_id, language.as_deref())
}

/// Run a preset's full adoption flow: create optional group, create team
/// shell, adopt each member template, bind to group, add to team, wire
/// connections. See `engine::team_preset_adopter` for the full step
/// sequence and partial-success semantics.
///
/// `language` is the user's current locale at the moment the adopt
/// button was clicked — passed straight through to the loader so the
/// persisted team + group names match what the user saw in the preview
/// modal. Switching language AFTER adoption does not retranslate
/// existing teams (they're frozen at the adopted locale).
///
/// `parameter_overrides` carries the combined questionnaire answers
/// from the preview modal: outer key is the preset-manifest role,
/// inner key is the question id, value is the answer (any JSON
/// type; downstream code coerces per the question's declared
/// `type`). Omit / pass `None` to adopt with template defaults.
///
/// Emits `team-preset-adopt-progress` events per member transition so
/// the preview modal can render a per-row status table. Returns
/// `AdoptedTeamPresetResult` with the new team_id, optional group_id,
/// successfully-adopted members, and any per-template failures.
#[tauri::command]
pub fn adopt_team_preset(
    state: State<'_, Arc<AppState>>,
    app: AppHandle,
    id: String,
    language: Option<String>,
    parameter_overrides: Option<
        std::collections::HashMap<
            String,
            std::collections::HashMap<String, serde_json::Value>,
        >,
    >,
) -> Result<AdoptedTeamPresetResult, AppError> {
    require_auth_sync(&state)?;
    team_preset_adopter::adopt_preset(
        &state,
        Some(app),
        &id,
        language.as_deref(),
        parameter_overrides.as_ref(),
    )
}

/// Retry the specified failed roles of a previously-adopted preset.
/// Surfaces as the "Retry N failed" button in `PresetPreviewModal` —
/// reuses the same `team-preset-adopt-progress` event stream so the
/// existing per-row status badges animate identically. Idempotent on
/// roles already adopted (silently skipped), so double-clicking is
/// safe.
///
/// `language` should match the locale used for the original adoption
/// so the retried members' persisted names stay consistent with the
/// rest of the team. (The frontend keeps the locale from the initial
/// adopt-click and passes the same value back here.)
///
/// Returns the FULL member list (existing + newly-retried) so the
/// modal can swap state in one assignment.
#[tauri::command]
pub fn retry_team_preset_members(
    state: State<'_, Arc<AppState>>,
    app: AppHandle,
    preset_id: String,
    team_id: String,
    group_id: Option<String>,
    roles: Vec<String>,
    language: Option<String>,
    parameter_overrides: Option<
        std::collections::HashMap<
            String,
            std::collections::HashMap<String, serde_json::Value>,
        >,
    >,
) -> Result<AdoptedTeamPresetResult, AppError> {
    require_auth_sync(&state)?;
    team_preset_adopter::retry_failed_members(
        &state,
        Some(app),
        &preset_id,
        &team_id,
        group_id.as_deref(),
        &roles,
        language.as_deref(),
        parameter_overrides.as_ref(),
    )
}
